File: packages/sdk/hush-go/guards/secret_leak.go
package guards

import (
	"fmt"
	"path/filepath"
	"regexp"

	"github.com/backbay/clawdstrike-go/internal"
	"github.com/backbay/clawdstrike-go/policy"
)

// DefaultSecretPatterns are the default secret detection patterns.
var DefaultSecretPatterns = []policy.SecretLeakPatternConfig{
	{Name: "aws_access_key", Pattern: `AKIA[0-9A-Z]{16}`, Severity: "critical"},
	{Name: "github_token", Pattern: `gh[ps]_[A-Za-z0-9]{36}`, Severity: "critical"},
	{Name: "openai_key", Pattern: `sk-[A-Za-z0-9]{48}`, Severity: "critical"},
	{Name: "private_key", Pattern: `-----BEGIN\s+(RSA\s+)?PRIVATE\s+KEY-----`, Severity: "critical"},
}

type compiledPattern struct {
	name     string
	re       *regexp.Regexp
	severity Severity
}

// SecretLeakGuard detects secrets in file writes.
type SecretLeakGuard struct {
	patterns  []compiledPattern
	skipPaths []string
}

// NewSecretLeakGuard creates a guard with the given config. Nil config uses defaults.
// Patterns are compiled at construction time; invalid patterns cause an error.
func NewSecretLeakGuard(cfg *policy.SecretLeakConfig) (*SecretLeakGuard, error) {
	rawPatterns := DefaultSecretPatterns
	var skipPaths []string

	if cfg != nil {
		if len(cfg.Patterns) > 0 {
			rawPatterns = cfg.Patterns
		}
		skipPaths = cfg.SkipPaths
	}

	compiled := make([]compiledPattern, 0, len(rawPatterns))
	for _, rp := range rawPatterns {
		re, err := regexp.Compile(rp.Pattern)
		if err != nil {
			return nil, fmt.Errorf("secret_leak: invalid pattern %q (%s): %w", rp.Name, rp.Pattern, err)
		}
		sev, _ := ParseSeverity(rp.Severity)
		compiled = append(compiled, compiledPattern{
			name:     rp.Name,
			re:       re,
			severity: sev,
		})
	}

	return &SecretLeakGuard{
		patterns:  compiled,
		skipPaths: skipPaths,
	}, nil
}

func (g *SecretLeakGuard) Name() string { return "secret_leak" }

func (g *SecretLeakGuard) Handles(action GuardAction) bool {
	return action.Type == "file_write" || action.Type == "patch"
}

func (g *SecretLeakGuard) Check(action GuardAction, ctx *GuardContext) GuardResult {
	// Normalize path before skip_paths check
	normalizedPath := filepath.Clean(action.Path)

	// Check skip_paths
	for _, skip := range g.skipPaths {
		if internal.DoubleStarMatch(skip, normalizedPath) {
			return Allow(g.Name())
		}
	}

	// For patch actions, scan the diff instead of content
	var content string
	if action.Type == "patch" {
		content = action.Diff
	} else {
		content = string(action.Content)
	}

	for _, pat := range g.patterns {
		match := pat.re.FindString(content)
		if match != "" {
			redacted := redact(match)
			return Block(g.Name(), pat.severity,
				fmt.Sprintf("potential secret detected (%s): %s", pat.name, redacted)).
				WithDetails(map[string]interface{}{
					"pattern": pat.name,
					"match":   redacted,
				})
		}
	}

	return Allow(g.Name())
}

// redact shows first 4 + "..." + last 4 characters.
func redact(s string) string {
	if len(s) <= 8 {
		return "****"
	}
	return s[:4] + "..." + s[len(s)-4:]
}

File: apps/academy/src/lib/guards/secret-leak.ts
/**
 * Secret Leak Guard Simulation
 *
 * Detects secrets in file writes using regex pattern matching.
 * Default patterns from rulesets/default.yaml.
 */
import type { GuardAction, GuardResult } from './types';

interface SecretPattern {
  name: string;
  pattern: string;
  severity: 'info' | 'warning' | 'error' | 'critical';
}

/** Default secret patterns from rulesets/default.yaml */
const DEFAULT_PATTERNS: SecretPattern[] = [
  {
    name: 'aws_access_key',
    pattern: 'AKIA[0-9A-Z]{16}',
    severity: 'critical',
  },
  {
    name: 'github_token',
    pattern: 'gh[ps]_[A-Za-z0-9]{36}',
    severity: 'critical',
  },
  {
    name: 'openai_key',
    pattern: 'sk-[A-Za-z0-9]{48}',
    severity: 'critical',
  },
  {
    name: 'private_key',
    pattern: '-----BEGIN\\s+(RSA\\s+)?PRIVATE\\s+KEY-----',
    severity: 'critical',
  },
];

/**
 * Compile a pattern string to a RegExp.
 * Handles (?i) prefix by converting to the 'i' flag.
 */
function compilePattern(pattern: string): RegExp {
  let flags = '';
  let src = pattern;
  if (src.startsWith('(?i)')) {
    flags = 'i';
    src = src.slice(4);
  }
  return new RegExp(src, flags);
}

export function evaluateSecretLeak(
  config: Record<string, unknown>,
  action: GuardAction,
): GuardResult {
  // Only evaluate file_write actions
  if (action.type !== 'file_write') {
    return {
      allowed: true,
      guard: 'secret_leak',
      severity: 'info',
      message: 'Action type not evaluated by secret_leak guard',
    };
  }

  const content = action.content;
  const configPatterns = config.patterns as
    | Array<{ name: string; pattern: string; severity?: string }>
    | undefined;

  const patterns: SecretPattern[] = configPatterns
    ? configPatterns.map((p) => ({
        name: p.name,
        pattern: p.pattern,
        severity: (p.severity as SecretPattern['severity']) ?? 'critical',
      }))
    : DEFAULT_PATTERNS;

  const matchedPatterns: string[] = [];
  let worstSeverity: SecretPattern['severity'] = 'info';
  const severityRank: Record<string, number> = {
    info: 0,
    warning: 1,
    error: 2,
    critical: 3,
  };

  for (const pat of patterns) {
    const regex = compilePattern(pat.pattern);
    if (regex.test(content)) {
      matchedPatterns.push(pat.name);
      if (severityRank[pat.severity] > severityRank[worstSeverity]) {
        worstSeverity = pat.severity;
      }
    }
  }

  if (matchedPatterns.length > 0) {
    return {
      allowed: false,
      guard: 'secret_leak',
      severity: worstSeverity,
      message: `Secret detected: ${matchedPatterns.join(', ')}`,
      details: {
        matched_patterns: matchedPatterns,
        path: action.path,
      },
    };
  }

  return {
    allowed: true,
    guard: 'secret_leak',
    severity: 'info',
    message: 'No secrets detected',
    details: { path: action.path },
  };
}
